musl: time64

xbjfk opened the pull request:
# Description
This change includes time64 support for applicable architectures (x86, arm, mips and powerpc). This is based on the previous PRs to this repo as well as the musl changelog from 1.1.24 -> 1.2. It can be enabled with the environment variable 
`RUST_LIBC_UNSTABLE_MUSL_TIME64` only when musl_v1_2_3 is enabled and the architecture is supported. 

A lot of structures, especially ones with mixed endian became excessively complicated, so I used `cfg_if` to separate them. It looks like you can only nest `s! {}` in `cfg_if! {}`, but not vice versa.

As a note, I'm considering removing `musl_not_time64`, and just keeping the old logic of allowing deprecated for function definitions involving `time_t` as it introduces necessary complexity.

When libc 1.0 is released, I believe the best path would be to remove the `musl_v1_2_3` feature, making it unconditionally enabled, keeping `musl_time64` which will expand to `(musl && time64_arch)`.

Tested through QEMU for all architectures.

# Sources
Sources are located on each commit, in the form of upstream commits

# Checklist

- [X] Relevant tests in `libc-test/semver` have been updated
- [X] No placeholder or unstable values like `*LAST` or `*MAX` are
  included (see [#3131](https://github.com/rust-lang/libc/issues/3131))
- [X] Tested locally (`cd libc-test && cargo test --target mytarget`);
  especially relevant for platforms that may not be checked in CI

@rustbot label +stable-nominated

xbjfk commented:
Since #4433 seems close to merging, I can rebase on top of it once merged.

xbjfk commented:
Rebased on top of the GNU changes :)

tgross35 reviewed (changes requested):
Thanks for working on this! Left a few comments here.

For the style check, are the changes it is suggesting possible? If not, that will probably have to be updated unfortunately (hopefully we will be able to reorganize things at some point so it's less annoying...)

xbjfk commented:
Thanks for the review! I'll get around to implementing the changes hopefully in the next few days.

martinetd commented on src/unix/mod.rs:
```diff
@@ -71,10 +71,14 @@ s! {
         #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
         pub tv_nsec: c_long,
+        #[cfg(all(musl_time64, target_endian = "little"))]
+        __pad0: u32,
```
Playing with this PR, this makes constructing timespec structs directly as was done in std fail:
```
error: cannot construct `timespec` with struct literal syntax due to private fields
   --> library/std/src/sys/pal/unix/thread.rs:251:30
    |
251 |                 let mut ts = libc::timespec {
    |                              ^^^^^^^^^^^^^^
    |
    = note: ...and other private field `__pad0` that was not provided
```

I'm not quite sure what to suggest here -- making these pub doesn't really make sense, and even if we did I believe we'd need something like `..Default::default()` to make it work so users of libc::timespec (including the compiler) will need to be fixed?... :/

At this build step there are only 4 (as of rust-lang/rust@255aa220821c05c3eac7605fce4ea1c9ab2cbdb4) so this is probably not too bad, but here goes my hopes to see this backported in 0.2 :p
(for reference:
 - library/std/src/sys/pal/unix/thread.rs:251:30
 - library/std/src/sys/pal/unix/time.rs:10:5
 - library/std/src/sys/pal/unix/time.rs:199:14
 - library/std/src/sys/fs/unix.rs:1507:24
)

martinetd replied:
FWIW, just reporting it's working as expected here but working around this issue by making the fields `pub` here and patching std as follow:
<details><summary>rust-lang/rust diff</summary>

```diff
diff --git a/Cargo.lock b/Cargo.lock
index 1a619096d34c..6f91779999af 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2046,8 +2046,6 @@ checksum = "9fa0e2a1fcbe2f6be6c42e342259976206b383122fc152e872795338b5a3f3a7"
 [[package]]
 name = "libc"
 version = "0.2.174"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "1171693293099992e19cddea4e8b849964e9846f4acee11b3948bcc337be8776"
 
 [[package]]
 name = "libdbus-sys"
diff --git a/Cargo.toml b/Cargo.toml
index c4d2a06f4cb1..12757101281e 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -89,3 +89,5 @@ codegen-units = 1
 # FIXME: LTO cannot be enabled for binaries in a workspace
 # <https://github.com/rust-lang/cargo/issues/9330>
 # lto = true
+[patch.crates-io.libc]
+path = "/mnt/rust/libc"
diff --git a/library/Cargo.lock b/library/Cargo.lock
index 34012d6943b7..b112d7ab9ac6 100644
--- a/library/Cargo.lock
+++ b/library/Cargo.lock
@@ -142,8 +142,6 @@ dependencies = [
 [[package]]
 name = "libc"
 version = "0.2.174"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "1171693293099992e19cddea4e8b849964e9846f4acee11b3948bcc337be8776"
 dependencies = [
  "rustc-std-workspace-core",
 ]
diff --git a/library/Cargo.toml b/library/Cargo.toml
index 35480b9319d7..955ae17cca35 100644
--- a/library/Cargo.toml
+++ b/library/Cargo.toml
@@ -51,3 +51,5 @@ rustc-std-workspace-core = { path = 'rustc-std-workspace-core' }
 rustc-std-workspace-alloc = { path = 'rustc-std-workspace-alloc' }
 rustc-std-workspace-std = { path = 'rustc-std-workspace-std' }
 compiler_builtins = { path = "compiler-builtins/compiler-builtins" }
+[patch.crates-io.libc]
+path = "/mnt/rust/libc"
diff --git a/library/std/src/sys/fs/unix.rs b/library/std/src/sys/fs/unix.rs
index dc278274f00f..db1fc8a527f0 100644
--- a/library/std/src/sys/fs/unix.rs
+++ b/library/std/src/sys/fs/unix.rs
@@ -1504,7 +1504,12 @@ pub fn set_times(&self, times: FileTimes) -> io::Result<()> {
                 io::ErrorKind::InvalidInput,
                 "timestamp is too small to set as a file time",
             )),
-            None => Ok(libc::timespec { tv_sec: 0, tv_nsec: libc::UTIME_OMIT as _ }),
+            None => Ok(libc::timespec {
+                tv_sec: 0,
+                tv_nsec: libc::UTIME_OMIT as _,
+                #[cfg(target_pointer_width = "32")]
+                __pad0: 0,
+            }),
         };
         cfg_if::cfg_if! {
             if #[cfg(any(target_os = "redox", target_os = "espidf", target_os = "horizon", target_os = "nuttx"))] {
diff --git a/library/std/src/sys/pal/unix/thread.rs b/library/std/src/sys/pal/unix/thread.rs
index d8b189413f4a..20002a4caf16 100644
--- a/library/std/src/sys/pal/unix/thread.rs
+++ b/library/std/src/sys/pal/unix/thread.rs
@@ -251,6 +251,8 @@ pub fn sleep(dur: Duration) {
                 let mut ts = libc::timespec {
                     tv_sec: cmp::min(libc::time_t::MAX as u64, secs) as libc::time_t,
                     tv_nsec: nsecs,
+                    #[cfg(target_pointer_width = "32")]
+                    __pad0: 0,
                 };
                 secs -= ts.tv_sec as u64;
                 let ts_ptr = &raw mut ts;
diff --git a/library/std/src/sys/pal/unix/time.rs b/library/std/src/sys/pal/unix/time.rs
index 0074d7674741..262f2d70fb18 100644
--- a/library/std/src/sys/pal/unix/time.rs
+++ b/library/std/src/sys/pal/unix/time.rs
@@ -6,8 +6,12 @@
 const NSEC_PER_SEC: u64 = 1_000_000_000;
 pub const UNIX_EPOCH: SystemTime = SystemTime { t: Timespec::zero() };
 #[allow(dead_code)] // Used for pthread condvar timeouts
-pub const TIMESPEC_MAX: libc::timespec =
-    libc::timespec { tv_sec: <libc::time_t>::MAX, tv_nsec: 1_000_000_000 - 1 };
+pub const TIMESPEC_MAX: libc::timespec = libc::timespec {
+    tv_sec: <libc::time_t>::MAX,
+    tv_nsec: 1_000_000_000 - 1,
+    #[cfg(target_pointer_width = "32")]
+    __pad0: 0,
+};
 
 // This additional constant is only used when calling
 // `libc::pthread_cond_timedwait`.
@@ -199,6 +203,8 @@ pub fn to_timespec(&self) -> Option<libc::timespec> {
         Some(libc::timespec {
             tv_sec: self.tv_sec.try_into().ok()?,
             tv_nsec: self.tv_nsec.as_inner().try_into().ok()?,
+            #[cfg(target_pointer_width = "32")]
+            __pad0: 0,
         })
     }
```
</details>

made our axum-based server run properly after changing the date past 2038.

(I'd be curious if there's a better way to test this though, perhaps with `-Zbuild-std` ?, but this is good enough for high level testing)

xbjfk replied:
Hmm, yeah - I'm not sure what the best path forward for this is. I believe the GNU changes will also result in a similar issue. I guess I will make the migration process to 1.0 more painful. I think I can make another PR to `rust` in that case. I think consumers will need to use `..Default::default()`. Another problem I can think of is people naively not including that when compiling for 64bit locally breaking things for time64 users.

xbjfk replied:
I'm not certain, but wouldn't `PhantomData` be fine if it is a member? I know ZST for FFI is generally not allowed - but the C code does not directly access the zero sized type? I can see that rust does not report `improper_ctypes` if a struct contains `PhantomData`.

martinetd replied:
My first reaction was that it makes sense to split the thorny part out for more discussion/refining, but thinking back it's probably wrong to use time64 variant functions with a mismatched type (as in will appear to work but fail subtly because ns are not where the callee thinks they are), so I think this has to stay together?
Even if the cost is that this never makes it to stable, I don't see how we could make a "partial time64 support" work properly.

I'll try to play some with struct shaping next week to see if we can do better than relying on default; I've never heard of anything that we could use to shape the struct without affecting the API but it can't hurt to spend a bit more time on it...

xbjfk replied:
Actually, I just realized - we don't need to use `PhantomData` hacks at all - we can just mark it as `#[non_exhaustive]` for all other architectures - this will have essentially the same effect, forcing people to use `..Default::default()` and such. Also, I think the idea is to fully enable time64 support and remove 32 bit time support, at least on `musl`.

You are right regarding the types - I meant to split making `timespec` `non_exhaustive` globally out.

xbjfk replied:
I created #4508

tgross35 replied:
In 1.0 we'll be making everything `non_exhaustive`, we more or less want everyone to go via `MaybeUninit::zeroed` and write individual fields like you have to do with C. Don't worry about it here while the flags are unstable.

See https://github.com/rust-lang/libc/issues/4080 for a bit more.

xbjfk commented:
Thanks for waiting - I've fixed most of the points, however I haven't done the style check just yet. It also looks like the `cfg!` change broke test compilation - there is a panic in `garando_syntax`.

xbjfk commented:
Okay, I have made a bit more progress. I ended up adding a whitelist for the "positive cfg" error - and whitelisted `musl32_time64` as well as `target_endian`.

The duplicate `s!` macro seems to be erroneously counting top-level/unconditional `s!` macros as the the same as `s!` macros within `cfg_if`. I might try find a way around this some other way if I can't figure out how to resolve the issue.

I have investigated the style crash when using `cfg!` and found that returning `Ok(None)` in `resolve_invoc` in `ctest` [here](https://github.com/rust-lang/libc/blob/90a377b3f48b0c8d29305c5173a9ffb052cbcd67/ctest/src/lib.rs#L1743) causes the issue. Always returning `Err(Determinacy::Determined)` seems to not crash. Looking at [garando](https://github.com/JohnTitor/garando/blob/5db92fce92f61fe0ace4011a39658ff82a2b58c3/garando_syntax/src/ext/expand.rs#L270), it seems to assume - `resolve_invoc` returns either `Err`, `Ok(...)` or be `InvocationKind::Attr`.

@alexcrichton Sorry for the ping but potentially you could help here? I know it's been a while since you wrote the `ctest` code referenced.

If we don't manage to resolve this, I think I will just revert back to separate statements, or maybe using `cfg_if` and adding a fixme.

alexcrichton commented:
Heh yes sorry at this point I'm no more an expert in ctest than anyone else!

xbjfk commented:
Pushed changes. Happy to wait for the ctest rewrite while we review :). I might need some help with the duplicate `s!` macro style checker, and how it marks `else` branches as duplicate with `s!` macros not in `cfg_if`.

xbjfk commented:
Rebased on top of main :)

xbjfk commented:
Hey @tgross35, sorry to ping/nag you - but I would be grateful if you could take a look at this :)

tgross35 commented on build.rs:
```diff
@@ -90,10 +93,15 @@ fn main() {
+        if musl32_time64 && MUSL_TIME64_ARCHS.contains(&target_arch.as_str()) {
+            set_cfg("musl32_time64");
+            set_cfg("linux_time_bits64");
+        }
```
(first commit) don't check for architectures explicitly, use `target_pointer_width` but opt out on specific arches if needed (similar to the GNU config)

tgross35 commented on build.rs:
```diff
@@ -24,6 +24,7 @@ const ALLOWED_CFGS: &[&str] = &[
     // Corresponds to `__USE_TIME_BITS64` in UAPI
     "linux_time_bits64",
     "musl_v1_2_3",
+    "musl32_time64",
```
Add a note about `_REDIR_TIME64` similar to the comments above

tgross35 commented on build.rs:
```diff
@@ -44,6 +45,8 @@ const CHECK_CFG_EXTRA: &[(&str, &[&str])] = &[
     ),
 ];
 
+const MUSL_TIME64_ARCHS: &[&str] = &["arm", "mips", "powerpc", "x86"];
```
Add a comment

`` /// Musl architectures that set `#define _REDIR_TIME64 1`. ``

I'd also call it `MUSL_REDIR_TIME64_ARCHES` to keep the lingo

tgross35 commented on build.rs:
```diff
@@ -90,10 +93,15 @@ fn main() {
     let musl_v1_2_3 = env::var("RUST_LIBC_UNSTABLE_MUSL_V1_2_3").is_ok();
     println!("cargo:rerun-if-env-changed=RUST_LIBC_UNSTABLE_MUSL_V1_2_3");
+    let musl32_time64 = env::var("RUST_LIBC_UNSTABLE_MUSL_TIME64").is_ok();
+    println!("cargo:rerun-if-env-changed=RUST_LIBC_UNSTABLE_MUSL_TIME64");
```
Do we actually need these to be distinct setting? If `musl_v1_2_3` is set then 64-bit time_t is implied. I don't think we really need to account for the 1.2.0..1.2.3 releases.

tgross35 commented on src/unix/linux_like/linux/musl/b32/arm/mod.rs:
```diff
@@ -39,12 +59,27 @@ s! {
+        #[cfg(musl32_time64)]
+        __st_ctim32: __c_anonymous_timespec32,
+        #[cfg(musl32_time64)]
+        pub st_ino: crate::ino_t,
```
I can't comment on the right lines, but isn't `struct stat64` identical here? If so, I think you could delete the struct and turn it into a type alias.

xbjfk replied:
Yeah - I checked the musl source code and I see that `stat64` is defined as `stat` - I'll double check each arch in musl and merge as necessary

xbjfk replied:
Yeah - I checked the musl source code and I see that `stat64` is defined as `stat` - I'll double check each arch in musl and merge as necessary. I noticed that __pad1 and __pad2 are pub in loongarch64, I'm not sure if making these private is acceptably breaking or not.

tgross35 commented on src/unix/linux_like/linux/mod.rs:
```diff
@@ -1,5 +1,8 @@
 //! Linux-specific definitions for linux-like values
 
+use core::cfg;
+use core::mem::size_of;
```
`size_of` should be in the prelude. Isn't `cfg` already available?

xbjfk replied:
Actually, cfg is not available already, it seems like it has to be explicit otherwise build fails..

tgross35 commented on src/unix/linux_like/linux/musl/b32/mips/mod.rs:
```diff
@@ -562,6 +562,13 @@ pub const SIGTSTP: c_int = 24;
+    note = "This signal actually corresponds to SIGEMT was incorrectly named in musl.
+        As it does not exist on mips it will be dropped entirely in a future release"
+)]
 pub const SIGSTKFLT: c_int = 7;
```
Just delete it, mips is tier3

tgross35 commented on ci/run-docker.sh:
```diff
@@ -45,6 +45,8 @@ run() {
         --env LIBC_CI_ZBUILD_STD \
         --env RUST_LIBC_UNSTABLE_GNU_FILE_OFFSET_BITS \
         --env RUST_LIBC_UNSTABLE_GNU_TIME_BITS \
+        --env RUST_LIBC_UNSTABLE_MUSL_V1_2_3 \
```
Also trying to fix this in https://github.com/rust-lang/libc/pull/4775

tgross35 commented:
Feel free to just add a commit deleting the `handle_s_macro_no_attrs` check, I don't have a good idea how to fix it to not complain in `cfg_if`

xbjfk commented on build.rs:
```diff
@@ -90,10 +93,15 @@ fn main() {
+            set_cfg("musl32_time64");
+            set_cfg("linux_time_bits64");
+        }
     }
```
Yeah, that's a lot cleaner - done!

xbjfk commented on src/unix/linux_like/linux/musl/b32/arm/mod.rs:
```diff
@@ -4,6 +4,11 @@ use crate::prelude::*;
 pub type wchar_t = u32;
 
 s! {
+    pub struct __c_anonymous_timespec32 {
```
done!

xbjfk commented on src/unix/linux_like/mod.rs:
```diff
@@ -13,6 +13,46 @@ missing! {
+                __reserved2: [c_long; 4],
+                //#[cfg(not(musl32_time64))]
+                //__reserved2: [crate::timespec; 2],
+                __reserved3: c_int,
```
Oh yeah, I just remembered why I did this - `musl_v1_2_3` will enable on `ohos` as well meaning an ABI break for OHOS. I made the changes but will leave this unresolved for now so we can confirm

xbjfk commented on libc-test/build.rs:
```diff
@@ -4826,7 +4826,9 @@ fn test_linux(target: &str) {
-        (old_musl && struct_ == "utmpx" && field == "ut_session")
+        (old_musl && struct_ == "utmpx" && field == "ut_session") ||
+        // FIXME(linux): this is changed to separate sec/usec fields when time64 is enabled
+        (struct_ == "input_event" && field == "time")
```
Actually, I don't think it is - not sure why test failures weren't shown on GNU, will need to investigate further..

xbjfk commented:
Okay, I've addressed most comments - unfortunately, it seems like there is an issue with compiling for hexagon - but it looks like it still appears on `main`.

tgross35 commented on src/unix/linux_like/mod.rs:
```diff
@@ -12,6 +12,53 @@ extern_ty! {
+            }
+        }
+    }
+}
```
Commit "musl: time64: update struct sched_param": to keep things a bit cleaner, move these musl definitions to the `musl` module. Then keep the existing `struct sched_param` in this module but add the deprecated notes and move it to a `cfg_if` block with the `any(musl, emscripten, ohos)` bit.

Also could you put this `cfg_if!` after the main `s!` macro with the others? Around line 252 looking at this file currently.

tgross35 commented on src/unix/linux_like/linux/mod.rs:
```diff
@@ -2786,9 +2787,9 @@ pub const GETNCNT: c_int = 14;
+pub const SEM_STAT: c_int = if cfg!(musl32_time64) { 0x112 } else { 18 };
 pub const SEM_INFO: c_int = 19;
-pub const SEM_STAT_ANY: c_int = 20;
+pub const SEM_STAT_ANY: c_int = if cfg!(musl32_time64) { 0x114 } else { 20 };
```
Ditto, may as well define them in terms of IPC_STAT https://github.com/bminor/musl/blob/1b76ff0767d01df72f692806ee5adee13c67ef88/include/sys/sem.h#L32-L34

tgross35 commented on src/unix/linux_like/linux/musl/b64/riscv64/mod.rs:
```diff
@@ -44,10 +44,10 @@ s! {
         pub st_size: off64_t,
         pub st_blksize: crate::blksize_t,
-        pub __pad2: c_int,
+        __pad2: c_int,
```
Ditto here

tgross35 commented:
Sorry for the conflicts here - more padding changes in https://github.com/rust-lang/libc/pull/4862, and a build config update in https://github.com/rust-lang/libc/commit/5c4255f8259298ac6ad159d1dd3a27484844d454.

martinetd commented:
Great, thanks for confirming!

I did a build with this branch back in June and it worked fine (at least I didn't notice any weird behavior, and my server didn't break when manually setting the date past 2038), but I'll definitely make time to retest at various parts of the process (soonish on master as a couple of things changed since, and when it gets backported etc), feel free to ping me if you want testing on later PRs and you remember (no harm either way).
